UNIXFS.md

```diff
@@ -74,15 +94,235 @@ message UnixTime {
+
+The minimum valid `PBNode.Data` field for a directory is (pseudo-json): `{"Type":"Directory"}`, other values are covered in Metadata.
+
+Every link in the Links list is an entry (children) of the directory, and the `PBNode.Links[].Name` field give you the name.
+
+####### Link ordering
+
+The cannonical sorting order is lexicographical over the names.
+
+In theory there is no reason an encoder couldn't use an other ordering, however this lose some of it's meaning when mapped into most file systems today (most file systems consider directories are unordered-key-value objects).
+
+A decoder SHOULD if it can, preserve the order of the original files in however it consume thoses names.
+
+However when some implementation decode, modify then reencode some, the orignal links order fully lose it's meaning. (given that there is no way to indicate which sorting was used originally)
+
+####### Path Resolution
+
+Pop the left most component of the path, and try to match it to one of the Name in Links.
+
+<!--TODO: check Kubo does this-->If you find a match you can then remember the CID. You MUST continue your search, however if you find a match again you MUST error.
+
+Assuming no errors were raised, you can continue to the path resolution on the mainaing component and on the CID you poped.
+
+####### Duplicate names
+
+Duplicate names are not allowed, if two identical names are present in an directory, the decoder MUST error.
 
-For files that are comprised of more than a single block, the 'Type' field will be set to 'File', the 'filesize' field will be set to the total number of bytes in the file (not the graph structure) represented by this node, and 'blocksizes' will contain a list of the filesizes of each child node.
+###### `Symlink` type
 
-This data is serialized and placed inside the 'Data' field of the outer merkledag protobuf, which also contains the actual links to the child nodes of this object.
+<!--TODO: check that this is true-->Symlinks MUST NOT have childs.
 
-For files comprised of a single block, the 'Type' field will be set to 'File', 'filesize' will be set to the total number of bytes in the file and the file data will be stored in the 'Data' field.
+Their Data.Data field is a POSIX path that maybe appended in front of the currently remaining path component stack.
 
-## Metadata
+####### Path resolution on symlinks
+
+There is no current consensus on how pathing over symlinks should behave.
+Some implementations return symlinks objects and fail if a consumer tries to follow it through.
+
+Following the POSIX spec over the current unixfs path context is probably fine.
+
+###### `HAMTDirectory`
+
+Thoses nodes are also sometimes called sharded directories, they allow to split directories into many blocks when they are so big that they don't fit into one single block anymore.
+
+- `node.Data.hashType` indicates a multihash function to use to digest path components used for sharding.
+It MUST be murmur3-x64-64 (multihash `0x22`).
```

Comment:
Do we want to have backward compatibility? It would seem the most prominent implementations are only taking the first 64 bits of the digest into account.
Maybe replace "lowest" in step 2 down on line 259 with a detailed description of how to pull out the correct bits from the middle of the 128-bit digest, in which case you could be backward-compatible with existing data and still allow the longer digest going forward (for large fanout and/or many, many, many entries).